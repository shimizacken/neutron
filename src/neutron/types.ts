export type Watcher<T> = (next?: T, previous?: T) => void;

export type Behavior = "default" | "re-emit";

export interface Neutron<T> {
  watch: (watcher: Watcher<T>) => () => boolean;
  abandon: (watcher: Watcher<T>) => boolean;
  emit: (nextState?: T) => void;
  getWatchers: () => Array<Watcher<T>>;
}
