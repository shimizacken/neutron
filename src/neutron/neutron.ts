import { Watcher, Neutron, Behavior } from "./types";

/**
 * Neutron observer
 */
export const neutron = <T>(previousState?: T) => (
  behavior: Behavior = "default"
): Neutron<T> => {
  const watchers = new Set<Watcher<T>>();
  /**
   * unsubscribes from current neutron
   * @param fnToRemove
   */
  const abandon = (watcherToRemove: Watcher<T>) =>
    watchers.delete(watcherToRemove);

  /**
   * subscribes to current neutron
   * @param watcher
   */
  const watch = (watcher: Watcher<T>) => {
    watchers.add(watcher);

    if (behavior === "re-emit for new watcher" && watchers.size > 1) {
      emitToSingleWatcher(watcher, previousState);
    }

    return () => abandon(watcher);
  };

  const emitToSingleWatcher = (watcher: Watcher<T>, nextState?: T) =>
    watcher(nextState);

  /**
   * fires new data to all observers
   * @param data
   */
  const emit = (next?: T) => {
    watchers.forEach((fn) => fn(next, previousState));

    previousState = next;
  };

  /**
   * return array of current watchers
   */
  const getWatchers = () => Array.from(watchers);

  return {
    watch,
    abandon,
    emit,
    getWatchers,
  };
};

/**
 * create a Neutron observer
 */
export const createNeutron = <T>(behavior?: Behavior): Neutron<T> =>
  neutron<T>()(behavior);
