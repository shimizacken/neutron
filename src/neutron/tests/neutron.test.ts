import { createNeutron } from "../neutron";

describe("neutron test suite", () => {
  it("should return", () => {
    const colorsNeutron = createNeutron<string>();

    expect(colorsNeutron).toMatchSnapshot();
  });
});
