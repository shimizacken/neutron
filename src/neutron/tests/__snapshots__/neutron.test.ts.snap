// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`neutron test suite should return 1`] = `
Object {
  "emit": [Function],
  "leave": [Function],
  "watch": [Function],
}
`;
